### PR TITLE
[BIOMAGE-2038] Improve numGenesVsNumUmis' responsiveness - UI

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Description

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-2038

#### Link to staging deployment URL (or set N/A)
https://ui-grasp6-sickly-worm.scp-staging.biomage.net/

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx [hunk 1]
```diff
@@ -1,24 +1,63 @@
...
+      {rerunRequired && (
+        <Space direction='vertical'>
+          <Alert
+            message='A new interactive version of this plot is available! To see it,
+      you need to run data processing again by clicking on the button below.
+      Please note that by doing so, any customisation made to your project will be lost.
+      In this case, we suggest creating a new project with the same data to view the new interactive plot.'
+            type='info'
+            showIcon
+          />
```
Comment (aerlaut) [inline src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx:50]: Do we really want the user to re-run their experiment and lose their existing data to see a new interactive plot?

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx [hunk 2]
```diff
@@ -1,24 +1,63 @@
...
+
+  // if prediction interval is not set - use the passed p-value from the pipeline
+  useEffect(() => {
+    if (!config.predictionInterval) {
+      updateSettings(
+        { predictionInterval: parseFloat(1 - config.regressionTypeSettings[config.regressionType]['p.level']).toFixed(6) },
+        false,
+      );
+    }
+  }, []);
```
Comment (aerlaut) [inline src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx]: This only run once, when the component is loaded, and it checks if a property `predictionInterval` is set in `config`, which by default does not exist. There is no guard in the plot which prevents this component from rendering before the `config` is loaded. Wouldn't this codeblock always run and `updateSettings` always called?

File: src/components/data-processing/GenesVsUMIs/GenesVsUMIs.jsx [hunk 3]
```diff
@@ -58,8 +58,6 @@ const GenesVsUMIs = (props) => {
 
   useEffect(() => {
     if (!config) {
-      const newConfig = _.clone(config);
-      _.merge(newConfig, expConfig);
```
Comment (aerlaut) [inline src/components/data-processing/GenesVsUMIs/GenesVsUMIs.jsx:62]: Why do you separate `expConfig` from `plotConfig`?

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx]: I think the update of this value should not be done here, but when generating plot spec.

File: src/components/data-processing/CalculationConfigContainer.jsx [hunk 4]
```diff
@@ -69,9 +69,11 @@ const CalculationConfigContainer = (props) => {
       {React.cloneElement(children, {
         config,
         plotType,
-        updateSettings: (diff) => {
+        updateSettings: (diff, outstandingChanges = true) => {
           dispatch(updateFilterSettings(filterUuid, diff, sampleId));
-          onFilterSettingsChange();
+          if (outstandingChanges) {
+            onFilterSettingsChange();
+          }
```
Comment (aerlaut) [inline src/components/data-processing/CalculationConfigContainer.jsx]: This code means that there are **setting changes** which does not cause **outstanding changes**, which is confusing. I see that you do this so that you can change the settings without causing the popup to showup. Is there a better way to do this?

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx (hunk 1 above)
Comment (gerbeldo) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx:50]: They can choose to re-run if they want the new interactivity, but if they don't want it or need it they can choose not to.  we could add something to this comment saying that.

Comment (gerbeldo) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx:50]:
```suggestion
      {/* only display info message for datasets which have
      not rerun the pipeline to see the new interractive plot */}
      {rerunRequired && (
        <Space direction='vertical'>
          <Alert
            message='A new interactive version of this plot is available! To see it,
      you need to run data processing again by clicking on the button below.
      Please note that by doing so, any customisation made to your project will be lost.
      In this case, we suggest creating a new project with the same data to view the new interactive plot. In case you do not need the interactivity, you can choose not to re-run data processing, and continue using the platform as it is.'
            type='info'
            showIcon
          />
```

Comment (StefanBabukov) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx:51]: Those were the instructions in the ticket. We are letting them know there is a new version in the plot.

File: src/components/data-processing/GenesVsUMIs/GenesVsUMIs.jsx (hunk 3 above)
Comment (StefanBabukov) [reply src/components/data-processing/GenesVsUMIs/GenesVsUMIs.jsx:62]: This was not doing anything so I removed it

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx (hunk 2 above)
Comment (StefanBabukov) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx]: good idea, I'll think of adding it there, we might even not need that other change checking for outstandingChanges that I also added for this.

**Comment by saracastel** [comment]
The y axis is moving a bit when moving the slider. In particular, the limit values change sometimes. For example, here is 4.0 with a prediction value of 0.8 and 4.2 with a prediction value of 0.99:
Screenshot 2022-08-01 at 13 41 49
Screenshot 2022-08-01 at 13 41 59
Maybe the limit Y-axis value could be calculated and set each time?

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx:50]: Can we simplify the wording? I think it's too long.

Comment (saracastel) [reply src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx:50]: > Can we simplify the wording? I think it's too long.

The message was already reviewed and approved by Vicky. I think it's worth to explain this in a proper way to the user, even if it's a bit longer. But if we come up with a shorter version, we can submit it to Vicky again.

**Comment by saracastel** [comment]
> Maybe the limit Y-axis value could be calculated and set each time?

Agreed with Vicky that this is a minor issue that can be fixed in a different ticket: https://biomage.atlassian.net/browse/BIOMAGE-2050?atlOrigin=eyJpIjoiNDE5ODY2M2YyNTNhNDM2OGIyYTRlMzkyM2E2MTQzNDMiLCJwIjoiaiJ9

File: src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx [hunk 5]
```diff
@@ -38,28 +74,66 @@ const GenesVsUMIsConfig = (props) => {
...
+
+        <Radio.Group
+          value={config.predictionInterval}
+          onChange={(val) => updateSettings({ predictionInterval: val.target.value })}
+          disabled={disabled}
+        >
+          <Space direction='vertical'>
+            <Space direction='horizontal'>
+              <InputNumber
+                value={newConfig.predictionInterval || 1 - config.regressionTypeSettings[config.regressionType]['p.level']}
```
Comment (aerlaut) [inline src/components/data-processing/GenesVsUMIs/CalculationConfig.jsx]: `1 - config.regressionTypeSettings[config.regressionType]['p.level'] ` is the default plevel? Can you make a new variable for this (e.g. `defaultPLevel`), and use that variable in here? It'll increase readibility

**Review by aerlaut — APPROVED** [review]

**Review by vickymorrison — APPROVED** [review]
